src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableRewriter.cs

```diff
@@ -255,25 +255,83 @@ protected override Expression VisitConstant(ConstantExpression c)
-        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2060:MakeGenericMethod",
-            Justification = "Enumerable methods don't have trim annotations.")]
         private static MethodInfo FindEnumerableMethodForQueryable(string name, ReadOnlyCollection<Expression> args, params Type[]? typeArgs)
         {
-            if (s_seqMethods == null)
+            s_seqMethods ??= GetEnumerableStaticMethods(typeof(Enumerable)).ToLookup(m => m.Name);
+
+            MethodInfo[] matchingMethods = s_seqMethods[name]
+                .Where(m => ArgsMatch(m, args, typeArgs))
+                .Select(ApplyTypeArgs)
+                .ToArray();
+
+            Debug.Assert(matchingMethods.Length > 0, "All static methods with arguments on Queryable have equivalents on Enumerable.");
+
+            if (matchingMethods.Length > 1)
             {
-                s_seqMethods = GetEnumerableStaticMethods(typeof(Enumerable)).ToLookup(m => m.Name);
+                return DisambiguateMatches(matchingMethods);
             }
-            MethodInfo? mi = s_seqMethods[name].FirstOrDefault(m => ArgsMatch(m, args, typeArgs));
-            Debug.Assert(mi != null, "All static methods with arguments on Queryable have equivalents on Enumerable.");
-            if (typeArgs != null)
-                return mi.MakeGenericMethod(typeArgs);
-            return mi;
+
+            return matchingMethods[0];
 
             [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2070:UnrecognizedReflectionPattern",
                 Justification = "This method is intentionally hiding the Enumerable type from the trimmer so it doesn't preserve all Enumerable's methods. " +
                 "This is safe because all Queryable methods have a DynamicDependency to the corresponding Enumerable method.")]
             static MethodInfo[] GetEnumerableStaticMethods(Type type) =>
                 type.GetMethods(BindingFlags.Public | BindingFlags.Static);
+
+            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2060:MakeGenericMethod",
+                Justification = "Enumerable methods don't have trim annotations.")]
+            MethodInfo ApplyTypeArgs(MethodInfo methodInfo) => typeArgs == null ? methodInfo : methodInfo.MakeGenericMethod(typeArgs);
+
+            // In certain cases, there might be ambiguities when resolving matching overloads, for example between
+            //   1. FirstOrDefault<object>(IEnumerable<object> source, Func<object, bool> predicate) and
+            //   2. FirstOrDefault<object>(IEnumerable<object> source, object defaultvalue).
+            // In such cases we disambiguate by picking a method with the most derived signature.
+            static MethodInfo DisambiguateMatches(MethodInfo[] matchingMethods)
+            {
+                Debug.Assert(matchingMethods.Length > 1);
+                ParameterInfo[][] parameters = matchingMethods.Select(m => m.GetParameters()).ToArray();
```

Comment:
I assume we don't need to worry about perf here? This could be written without using Linq.

Reply:
I suppose we could squeeze out a bit of extra performance, but I guess I simply mirrored convention in the particular class; we seem to be using Linq fairly liberally.